src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1147,9 +1147,25 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                 if (conn.TryGetBoundConnector(out var connector))
                 {
                     connector.StartUserAction(this);
+
+                    CancellationTokenSource? cts = null;
+
                     try
                     {
-                        using var _ = cancellationToken.Register(cmd => ((NpgsqlCommand)cmd!).Cancel(), this);
+                        if (cancellationToken.CanBeCanceled)
+                            cts = new CancellationTokenSource();
+
+                        using var _ = cancellationToken.Register(cmd =>
```

Comment:
We can avoid registering if the cancellation token can't be cancelled (which we check just above). See https://github.com/npgsql/npgsql/pull/2758.